pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaBase.hpp

```diff
@@ -0,0 +1,292 @@
+     * @param length Number of complex elements.
+     */
+    inline void CopyGpuDataToGpuIn(const CFP_t *gpu_sv, std::size_t length,
+                                   bool async = false) {
+        PL_ABORT_IF_NOT(BaseType::getLength() == length,
+                        "Sizes do not match for Host and GPU data");
+        data_buffer_->CopyGpuDataToGpu(gpu_sv, length, async);
+    }
+    /**
+     * @brief Explicitly copy data from another GPU device memory block to this
+     * GPU device.
+     *
+     * @param sv LightningGPU object to send data.
+     */
+    inline void CopyGpuDataToGpuIn(const Derived &sv, bool async = false) {
+        PL_ABORT_IF_NOT(BaseType::getNumQubits() == sv.getNumQubits(),
+                        "Sizes do not match for Host and GPU data");
+        auto same =
+            std::is_same_v<typename std::decay_t<typename std::remove_pointer_t<
+                               decltype(data_buffer_->getData())>>,
+                           typename std::decay_t<typename std::remove_pointer_t<
+                               decltype(sv.getData())>>>;
+        PL_ABORT_IF_NOT(same,
+                        "Data types are incompatible for GPU-GPU transfer");
+        data_buffer_->CopyGpuDataToGpu(sv.getData(), sv.getLength(), async);
+    }
+
+    /**
+     * @brief Explicitly copy data from host memory to GPU device.
+     *
+     * @param host_sv Complex data pointer to array.
+     * @param length Number of complex elements.
+     */
+    inline void CopyHostDataToGpu(const std::complex<Precision> *host_sv,
+                                  std::size_t length, bool async = false) {
+        PL_ABORT_IF_NOT(BaseType::getLength() == length,
+                        "Sizes do not match for Host and GPU data");
+        data_buffer_->CopyHostDataToGpu(
+            reinterpret_cast<const CFP_t *>(host_sv), length, async);
+    }
+
+    /**
+     * @brief Explicitly copy data from GPU device to host memory.
+     *
+     * @param sv Complex data pointer to receive data from device.
+     */
+    inline void CopyGpuDataToHost(std::complex<Precision> *host_sv,
+                                  size_t length, bool async = false) const {
+        PL_ABORT_IF_NOT(BaseType::getLength() == length,
+                        "Sizes do not match for Host and GPU data");
+        data_buffer_->CopyGpuDataToHost(host_sv, length, async);
+    }
+
+    /**
+     * @brief Explicitly copy data from this GPU device to another GPU device
+     * memory block.
+     *
+     * @param sv LightningGPU object to receive data.
+     */
+    inline void CopyGpuDataToGpuOut(Derived &sv, bool async = false) {
+        PL_ABORT_IF_NOT(BaseType::getNumQubits() == sv.getNumQubits(),
+                        "Sizes do not match for GPU data objects");
+        sv.getDataBuffer()->CopyGpuDataToGpu(getData(),
+                                             data_buffer_->getLength(), async);
+    }
```

Comment:
Nothing to do for now with this PR, but I think we can later move these functions out and call them from the databuffer only to clean-up the API.